readme.md
[![Build Status](https://travis-ci.com/k-box/k-search.svg?branch=develop)](https://travis-ci.com/k-box/k-search)

# K-Search

The K-Search is the interface of the search component for the K-Link project. It consists of a set of APIs
that enable third party clients to:

1. send data to be analyzed by the K-Search internal engine and 
2. perform a full-text search over such data.


## Installation

K-Search can be installed on most operating systems. The setup is heavily based on [Docker](https://www.docker.com/).

### Prerequisites

- Check the [system requirements](./docs/requirements.md).
- Use an operating system [supported by Docker](https://docs.docker.com/install/#server) (we use [Debian](https://debian.org) in production)
- Make sure you have installed the latest version of [Docker](https://docs.docker.com/install/linux/docker-ce/debian/) and [Docker Compose](https://docs.docker.com/compose/install/).

### Simplest installation

These few commands allow you to quickly install a K-Search **locally** on your computer for testing purposes.

* Create a directory: `mkdir k-search && cd k-search`
* Download configuration file: `curl -o docker-compose.yml https://raw.githubusercontent.com/k-box/k-search/develop/docker-compose.yml.dist`
* Start up services: `docker-compose up -d` (when running this for the first time, it will download a lot of data and take a while)
* Visit your K-Search: [http://localhost:8080/docs](http://localhost:8080/docs) (no login is required in the default setup).

## Usage

The general documentation is located under the [`docs` folder](./docs). 

Here is a brief table of contents for the most important parts:

- [HTTP API definition](./docs/api.md).
- [Framework Architecture](./docs/framework-architecture.md)
- [CLI Commands](./docs/commands.md)

## Development

The K-Search is implemented in PHP (v7.1.x) by using the [Symfony](http://symfony.com/)
framework.
The project structure follows the latest Symfony (v4.x) recommendation, as available at
[Symfony Architecture page](http://symfony.com/doc/current/quick_tour/the_architecture.html).

### Setup using Vagrant

This project uses [Drifter](https://github.com/liip/drifter) to provide a replicable development
environment, where most of the needed configurations and external services are installed and
configured.

The only requirement is a working installation of [Vagrant](https://www.vagrantup.com/).
Drifter supports both VirtualBox and the LXC containers, choose the one that best suits your needs.
Refer to the [Drifter documentation](https://liip-drifter.readthedocs.io/en/stable/requirements.html) 
for further details about its requirements and configuration.

To create working instance of a VM for the K-Search development and testing run the following
commands:

1. Obtain the K-Search code: `git clone https://github.com/k-box/k-search.git --recursive`
2. Start the VM and let Drifter provision it: `vagrant up`

The K-Search VM will be available at `http://ksearch.test`.

### Dependencies

The K-Search uses a set of Open Source technologies to support the full-text search and
data handling/manipulation functionality:

- [K-Search Engine](https://github.com/k-box/k-search-engine), based on [Apache Solr](http://lucene.apache.org/solr/): Search engine for data indexing and searching;
- [Apache Tika](https://tika.apache.org/): Text and metadata extraction from multiple file formats, bundled with Solr.



## Contributing

Contributions to the K-Search API are accepted by opening a Pull Request on the project page on GitHub.

1. the code must conform to the [Symfony Coding Standards](https://symfony.com/doc/current/contributing/code/standards.html);
2. the code must be covered by tests (`PHPUnit`, ...);
3. a changelog is provided with a summary of the changes.

The K-Search API code is regularly checked with the following tools:

- `php-cs-fixer`: to comply to the Symfony Coding Standards and additional PSRs (see: http://cs.sensiolabs.org)
- `phpstan`: Static analysis checks for PHP with a check-level of `7` (see: https://github.com/phpstan/phpstan)  
- `phpunit`: PHP testing framework (see: https://phpunit.de/)

## License

This project is licensed under the AGPL v3 license, see [LICENSE.txt](./LICENSE.txt).


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).

**Versioning schema**

The code version follows [Semantic Versioning](http://semver.org/), while the release version includes
a package number to track changes to the packaging: `{Major}.{Minor}.{Patch}-{package version}`.

The `Major`, `Minor` and `Patch` placeholders are defined as described in the Semantic Versioning
documentation, and:

 - the `{package version}` is a number that will start from `1` and increment with every changes made
   to the Docker image or repository structure.
 - The `{package version}` number is set to `1` when a new `{Major}.{Minor}.{Patch}` change is
   released (e.g. `3.0.1-1`, `3.0.1-2`, `3.0.2-1`, `3.1.0-1`).

## [Unreleased]

### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

## [3.7.0-1-beta2] - 2019-04-15

### Changed

- Filter on all K-Links defined for the application if no K-Link is specified

## [3.7.0-1-beta1] - 2019-04-03

### Added

- Support for multiple K-Links (API 3.7) _only available with K-Link Registry v0.3_
 - Ability to filter based on K-Link
 - `klink.list` endpoint to list available K-Links


## [3.6.0-2] - 2018-12-04

### Fixed

- Docker container shutdown

## [3.6.0-1] - 2018-10-31

### Changed

- For respecting EU 679/2016 (GDPR) we have removed the required constraints from the following `Data` fields:
 - `authors` can now be null or an empty array
 - `copyright` can now be null. If not null both properties (`owner` and `usage`) can be null
 - `uploader` is not anymore a required field. If specified the `name` attribute is optional
- `uploader.email` data attribute can now be null in respect of EU 679/2016 (GDPR)
- Changes to required fields are marked with API 3.6. To comply with the regulation the changes have been 
  backported to previous API versions

## [3.5.1-1] - 2018-10-24

### Fixed

- Fix validation of search parameters against null values
- Fix validation of `geo_location` and `geo_location_filter` for WGS84 coordinate system (handle LUCENE-8522 bug)

## [3.5.0-1] - 2018-09-21
### Added
- `geo_location` field in the `Data` object for storing geographic information related to the data (API `3.5`)
- `geo_location_filter` to the `SearchQuery` object for filtering the data based on the `Data.geo_location` field (API `3.5`)

### Changed
- Updated Solr to v7.4.x
- Now Requires K-Search Engine v1.x

## [3.4.0-1] - 2018-07-27
### Added
- Download Data files when data has been added with `data_textual_contents` (and `KLINK_RETAIN_DOWNLOADED_CONTENTS=1`).
### Changed
- Enforce validation for the `request.id` property in JsonRPC requests

## [3.3.0-2] - 2018-07-16
### Fix
- Fixed Docker integration

## [3.3.0-1] - 2018-06-22
### Added
- Add `ksearch:data:compute-hash` command to compute the expected hash of a file
- Add `var/data.db` Database to handle Data processing queue
- Allow `data.status` to return status form the index, or from the "processing queue" state (API `v3.4`)
- Allow stored files to be served from `/files/{UUID}`
- Allow "data_textual_contents" from data.add to be retained (env variable: KLINK_RETAIN_DOWNLOADED_CONTENTS=[1|0])
- Add configuration for downloaded file retention policy (env variable: KLINK_RETAIN_DOWNLOADED_CONTENTS=[1|0])
- Add "uploader.organization" property (filterable, sortable, facetable)

### Changed
- Refactored data processing with advanced queue system (default to file, supports RabbitMQ and other message queue systems)
- Move Vagrant VM default hostname from `ksearch.dev` to `ksearch.test`
- Allow all data to be searched and get, ignoring the data.status value

### Deprecated
- Configuration of `KLINK_REGISTRY_API_URL` is deprecated when `KLINK_REGISTRY_ENABLED` is set to 0

### Fixed
- Fix Data.Search filter validation for groups
- Fix parsing of query sent in `data.search` "filters" params, handle validation and Solr errors
- Fix cache invalidation for downloaded files. The data.hash is now additionally check against the existing file

## [3.2.1-1] - 2018-03-20
### Changed
- Update Docker composer file, added docs and configurations
- Update Docker integration (ignore folders, use optimized autoloader during `composer install`)
- Enable `postbigrequest` plugin for SolariumPHP (use POST method for big requests against Solr APIs)

### Fixed
- Fix extraction of Java examples form PDF (updated solr-engine to v0.4.1)

## [3.2.0-1] - 2018-02-16
### Added
- Allow Data files to be pre-downloaded in `var/data-downloads/xx/UUID` (where `xx` are the first two chars of the
  Data UUID property). The file is checked for an indexable mime-type and kept after indexing.
  The file is deleted after the deletion of the related Data from the index.
- Add `filename` to the full-text search matching
- Add min-count parameter to search aggregations (API v3.2)

### Changed
- Update Solr submodule (use github repository, v0.4.0)
- Update Drifter submodule (use newest release with native SF4 support)

## [3.1.1-1] - 2018-01-19
### Added
- Add sorting of Data search results (API v3.1)

### Changed
- Update Solr submodule to v5.5.5
- Update Drifter to latest release (added support for Symfony4)
- Updated `CopyrightOwner`: set `name` as required, renamed `contact` to `address` (**breaking change**).


src/Model/Data/Data.php
<?php

namespace App\Model\Data;

use App\Validator\Constraints\ValidGeoLocation;
use JMS\Serializer\Annotation as JMS;
use Swagger\Annotations as SWG;
use Symfony\Component\Validator\Constraints as Assert;

/**
 * @SWG\Definition(
 *     definition="Data\Data",
 *     required={"uuid", "hash", "type", "url", "author", "copyright", "uploader", "properties"}
 * )
 */
class Data
{
    public const DATA_TYPE_DOCUMENT = 'document';
    public const DATA_TYPE_VIDEO = 'video';

    /**
     * The Universally unique identifier of this data.
     *
     * @var string
     * @Assert\NotBlank()
     * @Assert\Uuid()
     * @JMS\Type("string")
     * @SWG\Property(
     *     example="cc1bbc0b-20e8-4e1f-b894-fb067e81c5dd",
     * )
     */
    public $uuid;

    /**
     * The URI where the source data is stored and retrievable.
     *
     * @var string
     * @Assert\NotBlank()
     * @Assert\Url()
     * @JMS\Type("string")
     * @SWG\Property(
     *     example="http://publicliterature.org/pdf/advsh12.pdf",
     * )
     */
    public $url;

    /**
     * The SHA-2 hash of the Document contents (SHA-512, thus 128 Chars).
     *
     * @var string
     * @Assert\NotBlank()
     * @Assert\Length(min="128", max="128")
     * @JMS\Type("string")
     * @SWG\Property(
     *     example="d6f644b19812e97b5d871658d6d3400ecd4787faeb9b8990c1e7608288664be77257104a58d033bcf1a0e0945ff06468ebe53e2dff36e248424c7273117dac09",
     * )
     */
    public $hash;

    /**
     * The general type of the provided data.
     *
     * @var string
     * @Assert\NotBlank()
     * @JMS\Type("string")
     * @Assert\Choice(
     *     strict=true,
     *     choices={"document", "video"}
     * )
     * @SWG\Property(
     *     enum={"document", "video"},
     *     example="video"
     * )
     */
    public $type;

    /**
     * The general type of the provided data.
     *
     * @var Klink[]
     * @JMS\ReadOnly()
     * @JMS\Type("array<App\Model\Data\Klink>")
     * @JMS\Since("3.7")
     * @SWG\Property(
     *     x={"since-version":"3.7"},
     * )
     */
    public $klinks = [];

    /**
     * The identifiers of the K-Links to which this data is published.
     *
     * This data is for internal use only, not exposed by default.
     *
     * @var array
     * @JMS\Type("array")
     * @JMS\Groups({"details"})
     * @JMS\Since("3.7")
     */
    public $klink_ids = [];

    /**
     * The properties of the data.
     *
     * @var Properties
     * @Assert\NotBlank()
     * @Assert\Valid()
     * @JMS\Type("App\Model\Data\Properties")
     * @SWG\Property()
     */
    public $properties;

    /**
     * List of authors (multiple).
     *
     * @var Author[]
     * @Assert\Type("array")
     * @Assert\Valid()
     * @JMS\Type("array<App\Model\Data\Author>")
     * @SWG\Property()
     */
    public $authors;

    /**
     * Information on the copyright.
     *
     * @var Copyright
     * @Assert\Valid()
     * @JMS\Type("App\Model\Data\Copyright")
     * @SWG\Property()
     */
    public $copyright;

    /**
     * The originating source where the data has been uploaded or created.
     *
     * @var Uploader
     * @Assert\Valid()
     * @JMS\Type("App\Model\Data\Uploader")
     * @SWG\Property()
     */
    public $uploader;

    /**
     * The Geo location of the data, as an escaped GeoJson string.
     *
     * The coordinates must be in the WGS84 coordinate system.
     * The order of the coordinates must be longitude, latitude
     *
     * @var string
     * @JMS\Type("string")
     * @JMS\Since("3.5")
     * @ValidGeoLocation()
     * @SWG\Property(
     *     property="geo_location",
     *     example="{""type"": ""Point"", ""coordinates"": [100.0, 0.0] }",
     *     x={"since-version":"3.5"},
     * )
     */
    public $geoLocation;

    /**
     * The status of the data, internal use only, not exposed.
     *
     * @var string
     * @JMS\Groups({"details"})
     */
    public $status;

    /**
     * The error status of the data, internal use only, not exposed.
     *
     * @var string|null
     * @JMS\Groups({"details"})
     */
    public $errorStatus;

    /**
     * The requestId that triggered the data, internal use only, not exposed.
     *
     * @var string
     * @JMS\Groups({"details"})
     */
    public $requestId;

    /**
     * The datetime when the data was updated, internal use only, not exposed.
     *
     * @var \DateTimeInterface
     * @JMS\Groups({"details"})
     */
    public $updatedAt;
}
